snuba/datasets/dataset.py

```diff
@@ -28,58 +28,107 @@ def get_min_columns(self) -> Sequence[str]:
 
 class Dataset(object):
     """
-    A dataset represent one or multiple entities in the Snuba data model.
+    A dataset represents a data model we can run a Snuba Query on.
+    A data model provides an abstract schema (today it is a flat table,
+    soon it will be a graph of Entities).
+    The dataset (later the Entity) has access to multiple Storage objects,
+    each one represents a table/view on the DB we can query.
     The class is a facade to access the components used to write on the
     data model and to query the entities.
-    To query the data model, it provides a schema (for one table or for
-    multiple joined tables), query processing features and query exension
-    parsing features.
-    To write it CAN provide a TableWriter, which has a schema as well and
-    provides a way to stream input from different sources and write them to
-    Clickhouse.
+
+    The dataset is made of several Storage objects (later we will introduce
+    entities between Dataset and Storage). Each storage represent a table/view
+    we can query.
+    When processing a query, there are three main steps:
+    - dataset query processing. A series of QueryProcessors are applied to the
+      query before deciding which Storage to use. These processors are defined
+      by the dataset
+    - the Storage to run the query onto is selected. This is done by a
+      QueryStorageSelector which is provided by the dataset. From this point
+      the query processing is storage specific. [This step is temporary till we
+      do not have a Query Plan Builder]
+    - storage query processing. A second series of QueryProcessors are applied
+      to the query. These are defined by the storage.
+
+    The architecture of the Dataset is divided in two layers. The highest layer
+    provides the logic we use to deal with the data model. (writers, query processors,
+    storage selectors, etc.). The lowest layer incldues simple objects that define
+    the query itself (Query, Schema, RelationalSource). The lop layer object access and
+    manipulate the lower layer objects.
     """
 
     def __init__(
         self,
-        dataset_schemas: DatasetSchemas,
         *,
-        table_writer: Optional[TableWriter] = None,
+        storages: Sequence[Storage],
+        storage_selector: QueryStorageSelector,
+        abstract_column_set: ColumnSet,
+        writable_storage: Optional[TableStorage],
```

Comment:
This also seems like it's going to need to move in the near future, yeah?

Reply:
as soon as we introduce entities.